### PR TITLE
fix(compiler): can't pass inflight closure as parameter to `super()` call

### LINKED ISSUE #4925: Can't pass an inflight closure to a `super` ctor call
### I tried this:

```js
class Base {
  new(handler: inflight (): void) { }
}

class Derived extends Base {
  new() {
    super(inflight () => { });
  }
}

new Derived();
```

### This happened:

```
➜  winglibs git:(main) wing compile ~/Documents/test.main.w 
ERROR: There is already a Construct with name 'Derived' in $Root [Default]

hint: Every preflight object needs a unique identifier within its scope. You can assign one as shown:

> new cloud.Bucket() as "MyBucket";

For more information, see https://www.winglang.io/docs/concepts/application-tree

../../Documents/target/test.main.wsim.543615.tmp/.wing/preflight.js:11
       class Base extends $stdlib.std.Resource {
         constructor($scope, $id, handler) {
>>         super($scope, $id);
         }
         static _toInflightType(context) {
```

### I expected this:

To work

### Is there a workaround?

_No response_

### Component

Compiler

### Wing Version

_No response_

### Node.js Version

_No response_

### Platform(s)

_No response_

### Anything else?

_No response_

### Community Notes

* Please vote by adding a 👍 reaction to the issue to help us prioritize.
* If you are interested to work on this issue, please leave a comment.

### PR DESCRIPTION
**Pull request by yoav-steinberg**
fixes #4925

Now you can pass inflight closures and also any `new PreflightClass()` as parameters to a `super()` call in a ctor.
What I did is detect these cases and made sure the implicit scope passed to the construct ctor is `$scope` and not `this` which isn't available before `super()` is called.

## Checklist

- [x] Title matches [Winglang's style guide](https://www.winglang.io/contributing/start-here/pull_requests#how-are-pull-request-titles-formatted)
- [x] Description explains motivation and solution
- [x] Tests added (always)
- [ ] Docs updated (only required for features)
- [ ] Added `pr/e2e-full` label if this feature requires end-to-end testing

*By submitting this pull request, I confirm that my contribution is made under the terms of the [Wing Cloud Contribution License](https://github.com/winglang/wing/blob/main/CONTRIBUTION_LICENSE.md)*.

### CODE REVIEW

File: examples/tests/valid/inflight_closure_as_super_param.test.w [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+    this.f_base = f;
+  }
+}
+
+class Derived extends Base {
+  pub f: Foo;
+  new() {
+    super(inflight (): str => { 
+      return "boom!";
+    }, new Foo() as "in_root");
```
Comment (Chriscbr) [inline examples/tests/valid/inflight_closure_as_super_param.test.w:17]: For the case of arbitrary preflight classes, the issue of not being able to pass a newly-constructed class to super() is interesting because the normal construction pattern (where the child class receives the "this" scope) also isn't possible in TypeScript:

```ts
class MyConstruct extends Construct {
  constructor(scope: Construct, id: string, helper: Construct) {
    super(scope, id);
    // do something with helper...
  }
}

class MySuperConstruct extends MyConstruct {
  constructor(scope: Construct, id: string) {
    let c = new Construct(this, "Helper"); // error: 'super' must be called before accessing this
    super(scope, id, c);
  }
}
```

As you observed, it's possible to reuse the "scope" to make it work:

```ts
let c = new Construct(scope, "Helper");
```

But it's worth noting this technically changes the tree structure so the construct being created is now a sibling instead of a child. Not necessarily a big deal.

Having the compiler do this automatically sounds like a fair compromise given that I don't think there's any other workaround. Indeed, I think the only other way to access the implicit "$scope" argument in Winglang right now requires the `nodeof()` function right now - but `nodeof` expects a scope, so it doesn't really help:

```
class A {
  new(bucket: cloud.Bucket) {
  }
}

class B extends A {
  new() {
    super(new cloud.Bucket() in nodeof(this).scope!); // error: using this before "super" has been called
  }
}
```

I feel like it could be worth documenting in the language reference or [here](https://github.com/winglang/wing/blob/main/docs/docs/02-concepts/02-application-tree.md) that classes constructed inside `super()` statements will be siblings instead of children in the tree - what do you think?

**Review by Chriscbr — APPROVED** [review]
Nice fix! 🙌

File: examples/tests/valid/inflight_closure_as_super_param.test.w (hunk 1 above)
Comment (yoav-steinberg) [reply examples/tests/valid/inflight_closure_as_super_param.test.w:17]: :+1:  Agree. I see the docs only contain information about preflight class "naming" and not about "scoping", I'll add a section and including this edge case.

**Review by eladb — APPROVED** [review]
Sweet!
